File: Util/pretty.py
"""
Generic Pretty Printer based on:

Pretty Priting
Derek C. Oppen, 1979
https://www.cs.tufts.edu/~nr/cs257/archive/derek-oppen/prettyprinting.pdf


This version does away with all the space optimizations and
offline concerns of the paper resulting in a drastically
simpler algorithm.

The API consists of a single function `PrettyPrint`.
Which takes list of Tokens and returns a string with
formatted text.

A token may be one of:
* Str -- a verbsatim string
* Brk -- an optional line break
* Beg -- begin of a group
* End -- end of a group

Enhancements to Oppen:
* a new BreakType, FORCE_LINE_BREAK, makes nesting easier
* NoBreak is a break that should not be broken

For usage see `tests.py`


Copyright 2025 Robert Muth <robert@muth.org>
See LICENSE for more details.
"""
from typing import Union

import enum
import dataclasses

_INFINIE_WIDTH = 1000000


class BreakType(enum.Enum):
    # No breaks are needed (the block fits on the line).
    # This will be automatically selected.
    FITS = 1
    # Fill line as much as possible then issue a line break and
    # continue in the same fashion on the next line
    INCONSISTENT = 2
    # If a block does not fit on a line, put each subblock
    # on a new line.
    CONSISTENT = 3
    # NOTE: THIS IS AN ADDITION TO THE ORIGINAL PAPER
    # Put each subblock on a new line
    # The `offset` applies even to the first line.
    # So this can be used as a nesting construct.
    FORCE_LINE_BREAK = 4


@dataclasses.dataclass()
class Str:
    string: str


@dataclasses.dataclass()
class Brk:
    """
    Either a space of widths `num_spaces` or a line break with additional
    `offset` relative to the indentation of the enclosing block
    """
    # Number of spaces if no overflow. Besides 1 the primary other values
    # are 0 and infinity (for LineBreak)
    num_spaces: int = 1
    # indent for overﬂow lines
    offset: int = 0
    # NOTE: THIS IS AN ADDITION TO THE ORIGINAL PAPER
    # This is useful with BreakType CONSISTENT to not automatically
    # force line breaks at the beginning of a group
    nobreak: bool = False


def LineBreak(offset=0):
    # Break which is guaranteed to overflow and hence forces a line break
    return Brk(num_spaces=_INFINIE_WIDTH, offset=offset)


def NoBreak(num_spaces):
    return Brk(num_spaces, 0, True)


@dataclasses.dataclass()
class Beg:
    # how to handle Breaks nested between this `Begin`` and `End``
    break_type: BreakType
    # additional indent
    offset: int


@dataclasses.dataclass()
class End:
    pass


Token = Union[Str, Brk, Beg, End]


def _ComputeSizes(tokens: list[Token]):
    # Stack invariants:
    # only Begin and Break tokens can be on the stack
    # there are never two consecutive Break Tokens on the Stack
    # So the top of the stack is eiher [... Begin] or
    # [... Begin Break]
    scan_stack = []
    sizes = []
    total = 0  # corresponds to `rightotal` in the paper
    for n, token in enumerate(tokens):
        if isinstance(token, Beg):
            sizes.append(-total)
            scan_stack.append(n)
        elif isinstance(token, End):
            sizes.append(1)
            x = scan_stack.pop(-1)
            sizes[x] += total
            if isinstance(tokens[x], Brk):
                x = scan_stack.pop(-1)
                sizes[x] += total
        elif isinstance(token, Brk):
            sizes.append(-total)
            # "close out" the last Break if there is one
            z = scan_stack[-1]
            if isinstance(tokens[z], Brk):
                z = scan_stack.pop()
                sizes[z] += total
            scan_stack.append(n)
            total += token.num_spaces
        elif isinstance(token, Str):
            sizes.append(len(token.string))
            total += len(token.string)
        else:
            assert False, f"unknown token {token}"
    assert len(tokens) == len(sizes), f"{len(tokens)} {len(sizes)}"
    return sizes


def _UpdateSizeOfNoBreaks(tokens: list[Token], sizes: list[int]):
    # Update chains of NoBreaks from the right to have smaller sizes
    # This will result in NoBreak to more likely fit in the current line
    total = _INFINIE_WIDTH
    for i in reversed(range(len(tokens))):
        token: Token = tokens[i]
        if isinstance(token, Beg):
            if token.break_type == BreakType.FORCE_LINE_BREAK:
                total = _INFINIE_WIDTH
        elif isinstance(token, End):
            total = _INFINIE_WIDTH
        elif isinstance(token, Brk):
            if token.nobreak:
                if total < sizes[i]:
                    sizes[i] = total
                    total += token.num_spaces
                else:
                    total = sizes[i]
            else:
                total = 0
        elif isinstance(token, Str):
            total += sizes[i]
        else:
            assert False
    # Add to the break preceding a sequence of NoBreaks
    # this should prevent breaking at NoBreaks because
    # we break earlier at the preceding Break
    total = 0
    for i in reversed(range(len(tokens))):

        token: Token = tokens[i]
        if isinstance(token, Beg):
            if token.break_type == BreakType.FORCE_LINE_BREAK:
                total = 0
        elif isinstance(token, End):
            pass
        elif isinstance(token, Brk):
            total += token.num_spaces
            if not token.nobreak:
                if total > sizes[i]:
                    sizes[i] = total
                total = 0
        elif isinstance(token, Str):
            total += sizes[i]


class _Output:

    def __init__(self: "_Output", line_width: int):
        self.buffer: list[str] = []
        self.remaining = line_width  # how much space is left in current line
        self.line_width = line_width  # constant

    def append(self, s):
        self.buffer.append(s)

    def get_string(self):
        return "".join(self.buffer)

    def append_with_space_update(self, s):
        self.buffer.append(s)
        self.remaining -= len(s)

    def indent_with_space_update(self, num_spaces):
        self.append_with_space_update(" " * num_spaces)

    def fits_in_current_line(self, size: int) -> bool:
        return size <= self.remaining

    def set_offset_and_line_break(self, offset: int):
        self.remaining = offset
        self.buffer.append("\n")
        self.buffer.append(" " * (self.line_width - self.remaining))


@dataclasses.dataclass()
class _Entry:
    offset: int  # same measurement as `remaining`, i.e.`offset`== margin means zero indent
    break_type: BreakType


def _Render(tokens, sizes, output: _Output):
    print_stack: list[_Entry] = []

    for token, size in zip(tokens, sizes):
        if isinstance(token, Beg):
            if token.break_type == BreakType.FORCE_LINE_BREAK:
                if print_stack:
                    offset = print_stack[-1].offset
                    output.set_offset_and_line_break(offset - token.offset)
                else:
                    offset = output.line_width
                entry = _Entry(
                    output.remaining, token.break_type)
            elif output.fits_in_current_line(size):
                entry = _Entry(0, BreakType.FITS)
            else:
                bt = (BreakType.CONSISTENT
                      if token.break_type == BreakType.CONSISTENT
                      else BreakType.INCONSISTENT)
                entry = _Entry(
                    output.remaining - token.offset, bt)
            print_stack.append(entry)
        elif isinstance(token, End):
            print_stack.pop()
            # Note, not resetting the offset right away in the
            # non-FORCE_LINE_BREAK case is responsible for the
            # behavior characterized in the paper as:
            # "In particular, it effectively changes (dynamically)
            # each occurrence of ] (string) into (string) ]."
        elif isinstance(token, Brk):
            top = print_stack[-1]
            break_type = top.break_type
            offset = top.offset
            if token.nobreak and output.fits_in_current_line(size) or break_type == BreakType.FITS:
                output.indent_with_space_update(token.num_spaces)
            elif break_type in (BreakType.CONSISTENT, BreakType.FORCE_LINE_BREAK):
                output.set_offset_and_line_break(offset - token.offset)
            elif top.break_type == BreakType.INCONSISTENT:
                if output.fits_in_current_line(size):
                    output.indent_with_space_update(token.num_spaces)
                else:
                    output.set_offset_and_line_break(offset - token.offset)

        elif isinstance(token, Str):
            # TODO: handle long/multiline strings
            assert size == len(token.string)
            output.append_with_space_update(token.string)
        else:
            assert False, f"Unsupported token {token}"


def PrettyPrint(tokens: list[Token], line_width: int) -> str:
    # print(tokens)
    sizes: list[int] = _ComputeSizes(tokens)
    _UpdateSizeOfNoBreaks(tokens, sizes)
    #for t, s in zip(tokens, sizes):
    #    print(t, s)
    output = _Output(line_width)
    _Render(tokens, sizes, output)
    return output.get_string()

File: Util/pretty.cc
// (c) Robert Muth - see LICENSE for more info

#include "pretty.h"

#include <cassert>
#include <iostream>

namespace PP {

std::ostream& operator<<(std::ostream& out, BreakType bt) {
  switch (bt) {
    case BreakType::INVALID:
      out << "INVALID";
      break;
    case BreakType::FITS:
      out << "FITS";
      break;
    case BreakType::INCONSISTENT:
      out << "INCONSISTENT";
      break;
    case BreakType::CONSISTENT:
      out << "CONSISTENT";
      break;
    case BreakType::FORCE_LINE_BREAK:
      out << "FORCE_LINE_BREAK";
      break;
  }
  return out;
}

std::ostream& operator<<(std::ostream& out, const Token& t) {
  switch (t.type) {
    case TokenType::BEG:
      out << "Beg(break_type=" << t.beg.break_type
          << ", offset=" << t.beg.offset << ")";
      break;
    case TokenType::END:
      out << "End()";
      break;
    case TokenType::BRK:
      out << "Brk(num_spaces=" << t.brk.num_spaces
          << ", offset=" << t.brk.offset << ", nobreak=" << t.brk.nobreak
          << ")";
      break;
    case TokenType::STR:
      out << "Str(\"" << t.str << "\")";
      break;
    case TokenType::INVALID:
      out << "INVALID";
      break;
  }
  return out;
}

template <typename T>
T PopStack(std::vector<T>& stack) {
  T top = stack.back();
  stack.pop_back();
  return top;
}

std::vector<ssize_t> ComputeSizes(const std::vector<Token>& tokens) {
  std::vector<ssize_t> sizes;
  sizes.reserve(tokens.size());
  ssize_t total = 0;
  std::vector<size_t> scan_stack;
  size_t x;

  for (size_t n = 0; n < tokens.size(); n++) {
    const Token& token = tokens[n];
    switch (token.type) {
      case TokenType::BEG:
        sizes.push_back(-total);
        scan_stack.push_back(n);
        break;
      case TokenType::END:
        sizes.push_back(1);
        x = PopStack(scan_stack);
        sizes[x] += total;
        if (tokens[x].type == TokenType::BRK) {
          x = PopStack(scan_stack);
          sizes[x] += total;
        }
        break;
      case TokenType::BRK:
        sizes.push_back(-total);
        // maybe close sizes a preceeding BRK
        x = scan_stack.back();
        if (tokens[x].type == TokenType::BRK) {
          scan_stack.pop_back();
          sizes[x] += total;
        }
        scan_stack.push_back(n);
        total += token.brk.num_spaces;
        break;
      case TokenType::STR:
        sizes.push_back(token.str.size());
        total += token.str.size();
        break;
      default:
        assert(false);  // unreachable
        break;
    }
  }
  return sizes;
}

void UpdatesSizesForNoBreaks(const std::vector<Token>& tokens,
                             std::vector<ssize_t>& sizes) {
  ssize_t total = INFINITE_WIDTH;
  for (size_t j = 0; j < tokens.size(); j++) {
    size_t i = tokens.size() - 1 - j;
    const Token& token = tokens[i];
    switch (token.type) {
      case TokenType::BEG:
        if (token.beg.break_type == BreakType::FORCE_LINE_BREAK) {
          total = INFINITE_WIDTH;
        }
        break;
      case TokenType::END:
        total = INFINITE_WIDTH;
        break;
      case TokenType::BRK:
        if (token.brk.nobreak) {
          if (total < sizes[i]) {
            sizes[i] = total;
            total += token.brk.num_spaces;
          } else {
            total = sizes[i];
          }
        } else {
          total = 0;
        }
        break;
      case TokenType::STR:
        total += token.str.size();
        break;
      case TokenType::INVALID:
        assert(false);  // unreachable
        break;
    }
  }
  //
  total = 0;
  for (size_t j = 0; j < tokens.size(); j++) {
    size_t i = tokens.size() - 1 - j;
    const Token& token = tokens[i];
    switch (token.type) {
      case TokenType::BEG:
        if (token.beg.break_type == BreakType::FORCE_LINE_BREAK) {
          total = 0;
        }
        break;
      case TokenType::END:
        break;
      case TokenType::BRK:
        total += token.brk.num_spaces;
        if (!token.brk.nobreak) {
          if (total > sizes[i]) {
            sizes[i] = total;
          }
          total = 0;
        }
        break;
      case TokenType::STR:
        total += token.str.size();
        break;
      case TokenType::INVALID:
        assert(false);  // unreachable
        break;
    }
  }
}

class Output {
 public:
  Output(size_t line_width, size_t approx_output_length)
      : line_width_(line_width), remaining_(line_width) {
    buffer_.reserve(approx_output_length);
  }

  void Append(std::string_view str) { buffer_ += str; }

  void AppendWithSpaceUpdate(std::string_view str) {
    buffer_ += str;
    remaining_ -= str.size();
  }

  void IndentWithSpaceUpdate(size_t num_spaces) {
    for (size_t i = 0; i < num_spaces; i++) {
      buffer_ += ' ';
    }
    remaining_ -= num_spaces;
  }

  size_t LineWidth() { return line_width_; }

  size_t Remaining() { return remaining_; }

  bool FitsInCurrentLine(size_t size) { return size <= remaining_; }

  std::string Get() { return buffer_; }

  void SetOffsetAndLineBreak(size_t offset) {
    remaining_ = offset;
    buffer_ += '\n';
    size_t ci = line_width_ - remaining_;
    for (size_t i = 0; i < ci; i++) {
      buffer_ += ' ';
    }
  }

 private:
  size_t line_width_;
  size_t remaining_;
  std::string buffer_;
};

struct Entry {
  size_t offset;
  BreakType break_type;
};

void Render(const std::vector<Token>& tokens, const std::vector<ssize_t>& sizes,
            Output* output) {
  std::vector<Entry> stack;
  for (size_t i = 0; i < tokens.size(); i++) {
    const Token& token = tokens[i];
    const size_t size = sizes[i];
    switch (token.type) {
      case TokenType::BEG: {
        Entry entry;
        if (token.beg.break_type == BreakType::FORCE_LINE_BREAK) {
          size_t offset;
          if (!stack.empty()) {
            offset = stack.back().offset;
            output->SetOffsetAndLineBreak(offset - token.beg.offset);
          } else {
            offset = output->LineWidth();
          }
          entry = {output->Remaining(), token.beg.break_type};
        } else if (output->FitsInCurrentLine(size)) {
          entry = {0, BreakType::FITS};
        } else {
          entry = {
              output->Remaining() - token.beg.offset,
              token.beg.break_type == BreakType::CONSISTENT
                  ? BreakType::CONSISTENT
                  : BreakType::INCONSISTENT,
          };
        }
        stack.push_back(entry);
      } break;
      case TokenType::END:
        stack.pop_back();
        break;
      case TokenType::BRK: {
        const Entry& top = stack.back();
        BreakType break_type = top.break_type;
        size_t offset = top.offset;
        if ((token.brk.nobreak && output->FitsInCurrentLine(size)) ||
            break_type == BreakType::FITS) {
          output->IndentWithSpaceUpdate(token.brk.num_spaces);
        } else if (top.break_type == BreakType::CONSISTENT ||
                   top.break_type == BreakType::FORCE_LINE_BREAK) {
          output->SetOffsetAndLineBreak(offset - token.brk.offset);
        } else if (top.break_type == BreakType::INCONSISTENT) {
          if (output->FitsInCurrentLine(size)) {
            output->IndentWithSpaceUpdate(token.brk.num_spaces);
          } else {
            output->SetOffsetAndLineBreak(offset - token.brk.offset);
          }
        }
      } break;

      case TokenType::STR:
        output->AppendWithSpaceUpdate(token.str);
        break;
      case TokenType::INVALID:
        assert(false);  // unreachable
        break;
    }
  }
}

size_t ApproxOutputLength(const std::vector<Token>& tokens) {
  size_t total = 0;
  for (const Token& token : tokens) {
    switch (token.type) {
      case TokenType::BEG:
      case TokenType::END:
        break;
      case TokenType::STR:
        total += token.str.size();
        break;
      case TokenType::BRK:
        total += token.brk.num_spaces;
        break;
      default:
        assert(false);  // unreachable
        break;
    }
  }
  return total;
}

std::string PrettyPrint(const std::vector<Token>& tokens, size_t line_width) {
  std::vector<ssize_t> sizes = ComputeSizes(tokens);
  assert(sizes.size() == tokens.size());
  UpdatesSizesForNoBreaks(tokens, sizes);
#if 0
  for (size_t i = 0; i < tokens.size(); i++) {
    std::cout << tokens[i] << " " << sizes[i] << std::endl;
  }
#endif
  Output output(line_width, ApproxOutputLength(tokens));
  Render(tokens, sizes, &output);
  return output.Get();
}

}  // namespace PP
